### PR TITLE
mnist: init at 2018-11-16

### PR DESCRIPTION
**Pull request by CMCDragonkai**
###### Motivation for this change

Adding the MNIST dataset. This makes it really useful to use any ML applications that depends on this dataset, and tests can be ran using this smaller dataset.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
I added the version as the date today because the package doesn't have versions. Unless you count the paper 1999 as the "version".

**Comment by CMCDragonkai** [comment]
Related: https://github.com/NixOS/nixpkgs/issues/46922

**Comment by CMCDragonkai** [comment]
I'm also wondering if this is the right way of doing multi-source derivations. I've just copied the style from Ocaml derivations. I also wonder about `mv` vs `cp` operations in the install phase. The files are not meant to be unpacked, as MNIST readers expect the data as compressed.

**Comment by CMCDragonkai** [comment]
I've discovered another way to create this. It's actually more minimal. Because the current derivation leaves the "srcs" still in the `/nix/store`, and ends up copying the data.

The below just creates symlinks to the individual data.

```nix
  let
    srcs = {
      train-images = fetchurl {
        url = "http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz";
        sha256 = "029na81z5a1c9l1a8472dgshami6f2iixs3m2ji6ym6cffzwl3s4";
      };
      train-labels= fetchurl {
        url = "http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz";
        sha256 = "0p152200wwx0w65sqb65grb3v8ncjp230aykmvbbx2sm19556lim";
      };
      test-images= fetchurl {
        url = "http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz";
        sha256 = "1rn4vfigaxn2ms24bf4jwzzflgp3hvz0gksvb8j7j70w19xjqhld";
      };
      test-labels= fetchurl {
        url = "http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz";
        sha256 = "1imf0i194ndjxzxdx87zlgn728xx3p1qhq1ssbmnvv005vwn1bpp";
      };
    };
  in
    linkFarm
    "mnist-2018-11-16"
    [
      { name = srcs.train-images.name; path = srcs.train-images; }
      { name = srcs.train-labels.name; path = srcs.train-labels; }
      { name = srcs.test-images.name; path = srcs.test-images; }
      { name = srcs.test-labels.name; path = srcs.test-labels; }
    ]
```

However I suspect this does not fit the conventions because there is not `maintainer` information nor is there any `version` information. Is there something equivalently simple? Or should I just swap the `cp` for `ln`?

File: pkgs/data/machine-learning/mnist/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    train-labels= fetchurl {
+      url = "http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz";
+      sha256 = "0p152200wwx0w65sqb65grb3v8ncjp230aykmvbbx2sm19556lim";
+    };
+    test-images= fetchurl {
+      url = "http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz";
+      sha256 = "1rn4vfigaxn2ms24bf4jwzzflgp3hvz0gksvb8j7j70w19xjqhld";
+    };
+    test-labels= fetchurl {
+      url = "http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz";
```
Comment (c0bw3b) [inline pkgs/data/machine-learning/mnist/default.nix]: Those URL can be considered stable, right? I mean, those datasets are frozen now and won't change anytime soon.

File: pkgs/data/machine-learning/mnist/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    version = "2018-11-16";
+    installPhase = ''
+      mkdir -p "$out"
+      ln -s "${srcs.train-images}" "$out/${srcs.train-images.name}"
+      ln -s "${srcs.train-labels}" "$out/${srcs.train-labels.name}"
+      ln -s "${srcs.test-images}" "$out/${srcs.test-images.name}"
+      ln -s "${srcs.test-labels}" "$out/${srcs.test-labels.name}"
+    '';
+    phases = [ "installPhase" ];
+    meta = with stdenvNoCC.lib; {
```
Comment (c0bw3b) [inline pkgs/data/machine-learning/mnist/default.nix]: Maybe also add the homepage attribute to point to http://yann.lecun.com/exdb/mnist/index.html

File: pkgs/data/machine-learning/mnist/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      description = "The MNIST database (Modified National Institute of Standards and Technology database) is a large database of handwritten digits that is commonly used for training various image processing systems";
```
Comment (c0bw3b) [inline pkgs/data/machine-learning/mnist/default.nix]: That's quite long for a short description. Maybe stick to `A large a large database of handwritten digits commonly used for training various image processing systems` 
and add a long description to explain more and define the MNIST acronym

File: pkgs/data/machine-learning/mnist/default.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      license = licenses.cc-by-sa-30;
```
Comment (c0bw3b) [inline pkgs/data/machine-learning/mnist/default.nix]: Where is this licensing documented?

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -3752,6 +3752,8 @@ with pkgs;
 
   mmv = callPackage ../tools/misc/mmv { };
 
+  mnist = callPackage ../data/machine-learning/mnist { };
```
Comment (c0bw3b) [inline pkgs/top-level/all-packages.nix]: There is a `### DATA` section further down the file.

File: pkgs/data/machine-learning/mnist/default.nix (hunk 1 above)
Comment (CMCDragonkai) [reply pkgs/data/machine-learning/mnist/default.nix]: This is the official upstream URL for MNIST data. There's no other source.

File: pkgs/data/machine-learning/mnist/default.nix (hunk 4 above)
Comment (CMCDragonkai) [reply pkgs/data/machine-learning/mnist/default.nix]: This is the only source I could find it that mentions it.

> Yann LeCun (Courant Institute, NYU) and Corinna Cortes (Google Labs, New York) hold the copyright of MNIST dataset, which is a derivative work from original NIST datasets. MNIST dataset is made available under the terms of the Creative Commons Attribution-Share Alike 3.0 license.
> http://www.pymvpa.org/datadb/mnist.html

File: pkgs/data/machine-learning/mnist/default.nix (hunk 3 above)
Comment (CMCDragonkai) [reply pkgs/data/machine-learning/mnist/default.nix]: Done.

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (CMCDragonkai) [reply pkgs/top-level/all-packages.nix]: Done.

File: pkgs/data/machine-learning/mnist/default.nix (hunk 2 above)
Comment (CMCDragonkai) [reply pkgs/data/machine-learning/mnist/default.nix]: Done

**Comment by CMCDragonkai** [comment]
@c0bw3b What do you think of using `ln` vs `cp`?

**Comment by c0bw3b** [comment]
I think it's preferable to `ln` over `cp` since `fetchurl` is going to install the sources in the nix store.

The only other approach I can think of would be to use `fetchzip` instead of fetchurl and `install -D -t $out <theFlatFilesWeWant>` in your install phase?
You may need to add `unpackPhase` to you list of phases for this to work. I have not tested locally.
`migmix` font seems to be close enough to what I have in mind.

**Comment by CMCDragonkai** [comment]
The files should not be decompressed.

**Comment by c0bw3b** [comment]
Oh.. Then I guess your current derivation is the better approach.

File: pkgs/data/machine-learning/mnist/default.nix [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      url = "http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz";
+      sha256 = "1imf0i194ndjxzxdx87zlgn728xx3p1qhq1ssbmnvv005vwn1bpp";
+    };
+  };
+in
+  stdenvNoCC.mkDerivation rec {
+    name = "mnist-${version}";
+    version = "2018-11-16";
+    installPhase = ''
+      mkdir -p "$out"
```
Comment (c0bw3b) [inline pkgs/data/machine-learning/mnist/default.nix]: (nitpick) you don't really need to quote `$out` here

File: pkgs/data/machine-learning/mnist/default.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
+{ stdenvNoCC, fetchurl }:
+let
+  srcs = {
+    train-images = fetchurl {
+      url = "http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz";
+      sha256 = "029na81z5a1c9l1a8472dgshami6f2iixs3m2ji6ym6cffzwl3s4";
+    };
+    train-labels= fetchurl {
```
Comment (c0bw3b) [inline pkgs/data/machine-learning/mnist/default.nix]:
```suggestion
    train-labels = fetchurl {
```
(nitpick) for consistency with the rest of nixpkgs
Idem for the next 2 ones

File: pkgs/data/machine-learning/mnist/default.nix [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    test-labels = fetchurl {
+      url = "http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz";
+      sha256 = "1imf0i194ndjxzxdx87zlgn728xx3p1qhq1ssbmnvv005vwn1bpp";
+    };
+  };
+in
+  stdenvNoCC.mkDerivation rec {
+    name = "mnist-${version}";
+    version = "2018-11-16";
+    installPhase = ''
```
Comment (veprbl) [inline pkgs/data/machine-learning/mnist/default.nix]: Can just use `buildCommand = ...` and drop `phases` override.

**Comment by veprbl** [comment]
Would it make sense to mark it with `preferLocalBuild = true`?

File: pkgs/data/machine-learning/mnist/default.nix (hunk 8 above)
Comment (CMCDragonkai) [reply pkgs/data/machine-learning/mnist/default.nix]: I tried it and it seem complicated.

**Comment by CMCDragonkai** [comment]
@veprbl why would we do that when none of the data packages do that?

**Review by c0bw3b — APPROVED** [review]
Tested locally. LGTM.

**Pull request merged**